### PR TITLE
Improve obervability of workflow execution

### PR DESCRIPTION
**Pull request by oliwel**
# Description

Add new observer events and accessors to the workflow object to improve observability. 
Makes handling of the return code from action->execute more robust when used with multiple result states.

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# Checklist:

- [X] My code follows the style guidelines of this project, please see the [contribution guidelines](https://github.com/jonasbn/perl-workflow/blob/master/CONTRIBUTING.md).
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

You might think, that this is one crazy checklist, but it is just as much written for the maintainer of the involved software :-)

### CODE REVIEW

File: lib/Workflow.pm [hunk 1]
```diff
@@ -313,13 +333,20 @@ sub _execute_single_action {
...
             && $self->log->info("Saved workflow with possible new state ok");
-     }
+
+        $self->notify_observers( 'completed', { state => $old_state, action => $action_name });
+
+        if ( $old_state ne $new_state ) {
+            $self->notify_observers( 'state change', { from => $old_state, action => $action_name, to => $new_state } );
+        }
+
+        return $self->_get_workflow_state;;
```
Comment (ehuelsmann) [inline lib/Workflow.pm:347]:
```suggestion
        return $self->_get_workflow_state;
```

File: lib/Workflow.pm [hunk 2]
```diff
@@ -861,40 +880,60 @@ No additional parameters.
...
-Adds the parameters C<$old_state>, and C<$action_name>.
-C<$old_state> includes the state of the workflow before the action
-was executed, C<$action_name> is the action name that was executed.
+Receives a hashref as second parameter holding the keys C<state> and
+C<action>. C<$state> includes the state of the workflow before the action
+was executed, C<$action> is the action name that was executed.
 
 =item *
 
 B<state change> - Issued after a workflow is successfully executed,
```
Comment (ehuelsmann) [inline lib/Workflow.pm:925]:
```suggestion
B<state change> - Issued after a single action is successfully executed,
```

File: lib/Workflow.pm [hunk 3]
```diff
@@ -861,40 +880,60 @@ No additional parameters.
...
 saved and results in a state change. The event will not be fired if
 you executed an action that did not result in a state change.
 
-Adds the parameters C<$old_state>, and C<$action>.
-C<$old_state> includes the state of the workflow before the action
-was executed, C<$action> is the action name that was executed.
+Receives a hashref as second parameter. The key C<from> holds the name
+of the state before the transaction, C<action> is the name of the action
```
Comment (ehuelsmann) [inline lib/Workflow.pm:930]:
```suggestion
of the state before the action, C<action> is the name of the action
```

File: lib/Workflow/Factory.pm [hunk 4]
```diff
@@ -991,6 +992,9 @@ if everything executes successfully and roll back if something fails.
 Note that you need to manage any L<Workflow::Persister::DBI::ExtraData>
 transactions yourself.
 
+If anything goes well, will inform all observers with the event C<save>
```
Comment (ehuelsmann) [inline lib/Workflow/Factory.pm:995]:
```suggestion
If everything goes well, will inform all observers with the event C<save>
```

**Review by ehuelsmann — APPROVED** [review]

**Comment by melutovich** [comment]
https://github.com/perl-workflow/perl-workflow/issues/257

**Pull request merged**